src/SourceTerms/Leakage/CPU_Src_Leakage.cpp

```diff
@@ -0,0 +1,1058 @@
+      break; // case 3
+   } // switch ( Mode )
+
+
+// (2-2) prepare fluid data for the leakage scheme
+#  ifdef MHD
+   const real Emag = (real)0.5*(  SQR( B[MAGX] ) + SQR( B[MAGY] ) + SQR( B[MAGZ] )  );
+#  else
+   const real Emag = NULL_REAL;
+#  endif
+
+   const real Dens_Code = fluid[DENS];
+   const real Eint_Code = Hydro_Con2Eint( fluid[DENS], fluid[MOMX], fluid[MOMY], fluid[MOMZ], fluid[ENGY],
+                                          true, MinEint, Emag );
+#  ifdef YE
+   const real Ye = fluid[YE] / fluid[DENS];
+#  else
+   const real Ye = NULL_REAL;
+#  endif
+
+#  ifdef TEMP_IG
+   const real Temp_IG_Kelv = fluid[TEMP_IG];
+#  else
+   const real Temp_IG_Kelv = NULL_REAL;
+#  endif
+
+#  if ( NUC_TABLE_MODE == NUC_TABLE_MODE_TEMP )
+   const int  NTarget = 0;
+#  else
+   const int  NTarget = 1;
+#  endif
+         int  In_Int[NTarget+1];
+         real In_Flt[4], Out[NTarget+1];
+
+   In_Flt[0] = Dens_Code;
+   In_Flt[1] = Eint_Code;
+   In_Flt[2] = Ye;
+   In_Flt[3] = Temp_IG_Kelv;
+
+   In_Int[0] = NTarget;
+#  if ( NUC_TABLE_MODE == NUC_TABLE_MODE_ENGY )
+   In_Int[1] = NUC_VAR_IDX_EORT;
+#  endif
+
+#  ifdef __CUDACC__
+   EoS->General_FuncPtr( NUC_MODE_ENGY, Out, In_Flt, In_Int, EoS->AuxArrayDevPtr_Flt, EoS->AuxArrayDevPtr_Int, EoS->Table );
+#  else
+   EoS_General_CPUPtr  ( NUC_MODE_ENGY, Out, In_Flt, In_Int, EoS_AuxArray_Flt,        EoS_AuxArray_Int,        h_EoS_Table );
+#  endif
+
+   const real Temp_Kelv = Out[0];
+
+
+// (3) get the change rates in internal energy and Ye from the leakage scheme
+   const bool NuHeat     = SrcTerms->Leakage_NuHeat;
+   const real NuHeat_Fac = SrcTerms->Leakage_NuHeat_Fac;
+   const real Unit_D     = SrcTerms->Unit_D;
+   const real Unit_T     = SrcTerms->Unit_T;
+         real dEdt_CGS, dYedt, Lum[NType_Neutrino], Heat[NType_Neutrino], NetHeat[NType_Neutrino];
+
+   Src_Leakage_ComputeLeak( Dens_Code, Temp_Kelv, Ye, chi, tau, Heat_Flux, HeatE_Rms, HeatE_Ave,
+                            &dEdt_CGS, &dYedt, Lum, Heat, NetHeat, NuHeat, NuHeat_Fac, Unit_D, EoS );
+
+// the returned dEdt_CGS is in erg/g/s
+   real dEdt_Code  = dEdt_CGS * sEint2Code * Unit_T * Dens_Code;
+   real dYedt_Code = dYedt * Unit_T;
+
+// make sure the new Ye is not within 1% of the table boundary
+   const real dYe = dYedt * dt;
+
+   if (  ( Ye + dYe < YeMin )  ||  ( Ye + dYe > YeMax )  )
+   {
+      dEdt_Code  = (real)0.0;
+      dYedt_Code = (real)0.0;
+   }
+
+
+// (4) update the internal energy density and Ye
+   const real Eint_Update = Eint_Code + dEdt_Code * dt;
+   const real Ye_Update   = Ye + dYedt_Code * dt;
+
+   fluid[ENGY] = Hydro_ConEint2Etot( fluid[DENS], fluid[MOMX], fluid[MOMY], fluid[MOMZ], Eint_Update, Emag );
+#  ifdef YE
+   fluid[YE  ] = Ye_Update * fluid[DENS];
+#  endif
+
+#  ifdef DYEDT_NU
+   fluid[DEDT_NU ] = FABS( dEdt_Code  );
+   fluid[DYEDT_NU] = FABS( dYedt_Code );
+#  endif
+
+
+// (5) store the dEdt, luminosity, heating rate, and net heating rate
+   if ( RecMode )
+   {
```

Comment:
i guess you use fluid to store some information for later records, but I thought it just updated in Line 555 and Line 556. Can you overwrite it now?

Reply:
The "Record Mode" is enabled at the beginning of `Record_CCSN_Leakage()` ([here](https://github.com/hfhsieh/gamer-fork/blob/ccsn_leakage/src/TestProblem/Hydro/CCSN/Record_CCSN.cpp#L244)), and is disabled at the end of this function ([here](https://github.com/hfhsieh/gamer-fork/blob/ccsn_leakage/src/TestProblem/Hydro/CCSN/Record_CCSN.cpp#L381)).  In `Record_CCSN_Leakage()`, we do not need the `dYedt` in L556, and the `dEdt` in L555 is stored in `fluid[DENS]` instead. In addition, the input `fluid[]` is a copy in a local variable (see [here](https://github.com/hfhsieh/gamer-fork/blob/ccsn_leakage/src/TestProblem/Hydro/CCSN/Record_CCSN.cpp#L311)) and will not affect the original fluid data.